<a href="https://colab.research.google.com/github/Rockrs/Algorithm_DS/blob/master/NumpyCaseStudy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np

Case Study

In [ ]:
! head cric_data.tsv

	Sachin Tendulkar	Rahul Dravid	India
0	100	78	342
1	11	62	191
2	8	85	252
3	71	24	307
4	104	17	229
5	18	104	246
6	8	76	226
7	86	74	288
8	12	60	216




*   Find Median, mean, IQR for Sachin and Rahul and India
*   Find Sachin's Histogram of Sachin's Score with 10 bins
*   Find Mean of Sachin's Score Grouped by 25 matches
*   Find Mean of Sachin's Score where Sachin score century
*   Find Mean of Sachin Score where Rahul has scored less than 10
*   Find Mean of Sachin score based on which quartile India's score falls
*   For every match who has scored more Sachin or Rahul
*   How many more run Sachin score on average after having scored x runs
*  How many matches did Sachin take to score 1000, 2000, 3000 and so on










### Problem 1

In [ ]:
cric_data = np.loadtxt('cric_data.tsv', skiprows =1)

In [ ]:
cric_data.shape

(225, 4)

In [ ]:
cric_data = cric_data[:, 1:]

In [ ]:
cric_data.shape

(225, 3)

In [ ]:
cric_data_sachin = cric_data[:, 0]
cric_data_rahul = cric_data[:, 1]
cric_data_india = cric_data[:, 2]

In [ ]:
def stats(col):
  print('Mean:', np.mean(col))
  print('Median:', np.median(col))
  print('IQR:', np.percentile(col, 75) - np.percentile(col, 25))

In [ ]:
stats(cric_data_sachin)

Mean: 39.87555555555556
Median: 27.0
IQR: 57.0


In [ ]:
stats(cric_data_rahul)

Mean: 32.062222222222225
Median: 22.0
IQR: 46.0


In [ ]:
stats(cric_data_india)

Mean: 220.79555555555555
Median: 216.0
IQR: 98.0


Efficient Way to do Above Operations

In [ ]:
np.mean(cric_data, axis =0)

array([ 39.87555556,  32.06222222, 220.79555556])

In [ ]:
np.median(cric_data, axis=0)

array([ 27.,  22., 216.])

In [ ]:
np.percentile(cric_data, 75, axis=0)  - np.percentile(cric_data, 25, axis=0)

array([57., 46., 98.])

###Problem 2

In [ ]:
np.histogram(cric_data_sachin)

(array([99, 36, 28, 16, 11, 17,  8,  8,  1,  1]),
 array([  0. ,  18.6,  37.2,  55.8,  74.4,  93. , 111.6, 130.2, 148.8,
        167.4, 186. ]))

In [ ]:
cric_data_sachin

array([100.,  11.,   8.,  71., 104.,  18.,   8.,  86.,  12.,  85.,  18.,
         4.,   7.,  37.,  14.,   0.,   4.,   0.,  21.,   1.,  62.,   0.,
       138.,  38.,   2.,  46.,  65.,   0.,  39.,  48., 141.,  62.,  12.,
         1.,  41.,  11.,   3., 186.,  11.,  27.,  27.,  51.,  18.,  32.,
       146.,   5.,  45., 141.,  12.,  65.,  27.,   7.,  16.,   2.,  28.,
         6., 123., 120.,   7.,   3.,   0.,  81.,   2.,  54., 122.,   4.,
        14.,   0., 100.,  15.,   0.,  57.,  99.,  37.,  38.,  32.,  21.,
        32.,  40.,   0.,   5.,   8.,   5.,   0.,  50.,  30.,  37.,  89.,
         4.,  98.,  83.,  93.,   0.,  52., 152.,   1.,   8.,  93.,  45.,
        26.,   0.,   1.,   0.,  16.,  47.,  89.,   3.,   1.,  53.,  16.,
         0.,  81.,  14.,  78.,   6., 105., 122.,   9.,   8.,  28.,  35.,
        69.,  13.,  97.,  93.,   2.,  36.,  39.,   2.,  29.,  12.,  19.,
        34.,   2., 100.,  44.,  82.,   0.,  79.,   6.,   9.,   8.,  23.,
        93.,  35.,  63.,  74.,   8., 117.,  39.,  4

### Problem 3

In [ ]:
cric_data_sachin.shape

(225,)

In [ ]:
sachin_by_25s = cric_data_sachin.reshape((9, 25))

In [ ]:
np.mean(sachin_by_25s, axis = 1)

array([33.96, 49.4 , 38.48, 40.16, 39.36, 38.2 , 44.6 , 39.52, 35.2 ])

### Problem 4, 5

In [ ]:
cric_data_sachin >=100

array([ True, False, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False,  True, False, False, False, False,
       False, False, False,  True, False, False, False, False, False,
       False,  True, False, False, False, False, False, False,  True,
       False, False,  True, False, False, False, False, False, False,
       False, False,  True,  True, False, False, False, False, False,
       False,  True, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False,  True, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True,  True,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [ ]:
np.mean(cric_data_sachin[cric_data_sachin >= 100])

125.0

In [ ]:
np.mean(cric_data_sachin[cric_data_rahul < 10])

40.74285714285714

### Problem 6

In [ ]:
qrt = np.percentile(cric_data_india, [25, 50, 75, 100])

If India's Score is less than 175 then what is Sachin's average
If India's Score is less than 216 then what is Sachin's average

In [ ]:
india_first_qrt = np.percentile(cric_data_india, [25])

In [ ]:
india_first_qrt

array([175.])

In [ ]:
type(india_first_qrt)

numpy.ndarray

In [ ]:
cric_data_sachin > india_first_qrt

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [ ]:
temp = np.array([
          [
           [1],
           [2]
           ],
          [
           [3],
           [4]
           ]
])

In [ ]:
temp.shape

(2, 2, 1)

In [ ]:
temp2 = np.array([
         [1,2],
         [3,4]
])

In [ ]:
temp2.shape

(2, 2)

In [ ]:
temp3 = temp + temp2

In [ ]:
temp3.shape

(2, 2, 2)

In [ ]:
temp3

array([[[2, 3],
        [5, 6]],

       [[4, 5],
        [7, 8]]])

In [ ]:
india_all_qtr = np.percentile(cric_data_india, [25, 50, 75, 100])

In [ ]:
india_all_qtr

array([175., 216., 273., 499.])

In [ ]:
cric_data_india > india_all_qtr

ValueError: ignored

In [ ]:
india_all_qtr.shape

(4,)

In [ ]:
cric_data_india.shape

(225,)

In [ ]:
temp4 = np.array([1,2,3,4,5])

In [ ]:
temp4.shape

(5,)

In [ ]:
temp4.reshape((5,1))

array([[1],
       [2],
       [3],
       [4],
       [5]])

In [ ]:
india_all_qtr = india_all_qtr.reshape((4, 1))

In [ ]:
cric_data_sachin < india_all_qtr

array([[ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True, False,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
      

In [ ]:
temp5.shape

(4, 225)

In [ ]:
temp5

array([[False, False, False, False, False, False, False, False, False,
        False,  True, False, False, False,  True, False, False,  True,
        False, False,  True, False, False, False, False,  True,  True,
         True,  True,  True, False, False, False, False, False, False,
         True, False,  True, False, False, False, False, False, False,
        False, False, False, False,  True,  True, False, False, False,
         True, False, False, False, False,  True, False, False, False,
        False, False,  True, False, False, False,  True, False, False,
        False, False, False, False, False, False,  True, False,  True,
         True, False, False, False, False, False,  True, False, False,
        False, False,  True, False, False, False, False, False, False,
        False,  True,  True,  True, False, False, False, False, False,
        False, False,  True,  True, False, False, False, False, False,
        False, False, False, False, False,  True, False, False,  True,
      

In [ ]:
indices  = cric_data_india < india_all_qtr

In [ ]:
indices.shape

(4, 225)

In [ ]:
sachin_when_india_less_than_first_qtr = cric_data_sachin[indices[0,:]]

In [ ]:
sachin_when_india_less_than_first_qtr

array([18., 14.,  0., 62., 46., 65.,  0., 39., 48.,  3., 11., 65., 27.,
       28.,  3.,  4., 15., 40.,  5.,  8., 89.,  0.,  0.,  1.,  0.,  0.,
       81., 13.,  2., 36., 12., 19.,  0.,  6., 35.,  0., 44.,  3., 47.,
       17., 35., 33.,  7.,  9.,  2., 11., 17.,  1., 10.,  0., 23.,  1.,
        2., 25.,  0.])

In [ ]:
for i in range(4):
  print ("i", np.mean(cric_data_sachin[indices[i]]))

i 19.672727272727272
i 28.18018018018018
i 31.688622754491018
i 39.799107142857146


### Problem 8

In [ ]:
hasRahulScoredGreaterThanSachin = cric_data_rahul > cric_data_sachin

In [ ]:
np.sum(hasRahulScoredGreaterThanSachin) / cric_data_india.size

0.4711111111111111

### Problem 10

Take cumulative sum using np.cumsum and  then use histogram using custom bin sizes

In [ ]:
sachin_cum = np.cumsum(cric_data_sachin)

In [ ]:
sachin_cum

array([ 100.,  111.,  119.,  190.,  294.,  312.,  320.,  406.,  418.,
        503.,  521.,  525.,  532.,  569.,  583.,  583.,  587.,  587.,
        608.,  609.,  671.,  671.,  809.,  847.,  849.,  895.,  960.,
        960.,  999., 1047., 1188., 1250., 1262., 1263., 1304., 1315.,
       1318., 1504., 1515., 1542., 1569., 1620., 1638., 1670., 1816.,
       1821., 1866., 2007., 2019., 2084., 2111., 2118., 2134., 2136.,
       2164., 2170., 2293., 2413., 2420., 2423., 2423., 2504., 2506.,
       2560., 2682., 2686., 2700., 2700., 2800., 2815., 2815., 2872.,
       2971., 3008., 3046., 3078., 3099., 3131., 3171., 3171., 3176.,
       3184., 3189., 3189., 3239., 3269., 3306., 3395., 3399., 3497.,
       3580., 3673., 3673., 3725., 3877., 3878., 3886., 3979., 4024.,
       4050., 4050., 4051., 4051., 4067., 4114., 4203., 4206., 4207.,
       4260., 4276., 4276., 4357., 4371., 4449., 4455., 4560., 4682.,
       4691., 4699., 4727., 4762., 4831., 4844., 4941., 5034., 5036.,
       5072., 5111.,

In [ ]:
sachin_histogram  = np.histogram(sachin_cum, bins =np.arange(0, 10000, 1000))

In [ ]:
sachin_histogram

(array([29, 18, 26, 25, 26, 26, 23, 22, 30]),
 array([   0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000]))